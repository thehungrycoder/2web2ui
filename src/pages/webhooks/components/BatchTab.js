import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import _ from 'lodash';

import { getBatches } from 'src/actions/webhooks';
import PanelLoading from 'src/components/panelLoading/PanelLoading';

import { Button, Tag } from '@sparkpost/matchbox';
import { TableCollection, Empty } from 'src/components';
import { selectWebhookBatches } from 'src/selectors/webhooks';

const columns = [
  { label: 'Timestamp' },
  { label: 'Batch ID' },
  { label: 'Status' },
  { label: 'Attempts' },
  null
];

export class BatchTab extends Component {
  componentDidMount() {
    this.refreshBatches();
  }

  refreshBatches = () => {
    const { webhook, getBatches } = this.props;
    const { id, subaccount } = webhook;
    return getBatches({ id, subaccount });
  };

  getRowData = (batch) => {
    const { webhook, query } = this.props;

    return [
      batch.formatted_time,
      batch.batch_id,
      <Tag color={batch.status === 'Success' ? 'blue' : 'yellow'}>{batch.response_code}</Tag>,
      batch.attempts,
      <div style={{ textAlign: 'right' }}>
        <Button component={Link} size='small' to={`/webhooks/details/${webhook.id}/batches/${batch.batch_id}${query}`}>View Details</Button>
      </div>
    ];
  };

  render() {
    const { batches, batchesLoading } = this.props;

    if (batchesLoading) {
      return <PanelLoading />;
    }

    if (_.isEmpty(batches)) {
      return <Empty message='There are no batches for your webhook' />;
    }

    return (
      <TableCollection
        columns={columns}
        rows={batches}
        getRowData={this.getRowData}
        pagination={true}
      />
    );
  }
}

const mapStateToProps = (state) => ({
  batches: selectWebhookBatches(state),
  batchesLoading: state.webhooks.batchesLoading
});

export default connect(mapStateToProps, { getBatches })(BatchTab);
