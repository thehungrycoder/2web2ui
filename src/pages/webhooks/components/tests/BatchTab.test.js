import { shallow } from 'enzyme';
import React from 'react';

import { BatchTab } from '../BatchTab';

describe('Webhook Component: Batch Status Tab', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      getBatches: jest.fn(() => Promise.resolve()),
      webhook: {
        id: 'webhook-id'
      },
      batches: [
        {
          formatted_time: 'so-formatted',
          batch_id: '243423423423',
          status: 'p',
          attempts: 1,
          response_code: 200
        },
        {
          formatted_time: 'so-formatted-2',
          batch_id: '996969545',
          status: 'f',
          attempts: 4,
          response_code: 500
        }
      ],
      batchesLoading: false,
      showAlert: jest.fn()
    };

    wrapper = shallow(<BatchTab {...props} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render batch status tab with table data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should show loading component while refreshing data', () => {
    wrapper.setProps({ batchesLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty message when no batches exist', () => {
    wrapper.setProps({ batches: []});
    expect(wrapper).toMatchSnapshot();
  });

  it('should call getBatches in mounting', () => {
    const instance = wrapper.instance();
    expect(instance.props.getBatches).toHaveBeenCalledWith({ id: instance.props.webhook.id, subaccount: undefined });
  });

  it('should getBatches when refreshing', () => {
    const instance = wrapper.instance();
    instance.refreshBatches();
    expect(instance.props.getBatches).toHaveBeenCalledWith({ id: instance.props.webhook.id, subaccount: undefined });
  });

  it('should getBatches with a subaccount', () => {
    wrapper.setProps({ webhook: { id: 'id-with-sub', subaccount: 101 }});
    const instance = wrapper.instance();
    instance.refreshBatches();
    expect(instance.props.getBatches).toHaveBeenCalledWith({
      id: instance.props.webhook.id,
      subaccount: instance.props.webhook.subaccount
    });
  });
});
