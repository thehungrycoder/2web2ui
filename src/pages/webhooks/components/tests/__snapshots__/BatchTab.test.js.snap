// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook Component: Batch Status Tab should render batch status tab with table data 1`] = `
<TableCollection
  columns={
    Array [
      Object {
        "label": "Timestamp",
      },
      Object {
        "label": "Batch ID",
      },
      Object {
        "label": "Status",
      },
      Object {
        "label": "Attempts",
      },
      null,
    ]
  }
  defaultSortColumn={null}
  defaultSortDirection="asc"
  getRowData={[Function]}
  pagination={true}
  rows={
    Array [
      Object {
        "attempts": 1,
        "batch_id": "243423423423",
        "formatted_time": "so-formatted",
        "response_code": 200,
        "status": "Success",
      },
      Object {
        "attempts": 4,
        "batch_id": "996969545",
        "formatted_time": "so-formatted-2",
        "response_code": 500,
        "status": "f",
      },
    ]
  }
/>
`;

exports[`Webhook Component: Batch Status Tab should render table rows correctly 1`] = `
Array [
  Array [
    "so-formatted",
    "243423423423",
    <Tag
      color="blue"
    >
      200
    </Tag>,
    1,
    <div
      style={
        Object {
          "textAlign": "right",
        }
      }
    >
      <Button
        component={[Function]}
        size="small"
        to="/webhooks/details/webhook-id/batches/243423423423"
      >
        View Details
      </Button>
    </div>,
  ],
  Array [
    "so-formatted-2",
    "996969545",
    <Tag
      color="yellow"
    >
      500
    </Tag>,
    4,
    <div
      style={
        Object {
          "textAlign": "right",
        }
      }
    >
      <Button
        component={[Function]}
        size="small"
        to="/webhooks/details/webhook-id/batches/996969545"
      >
        View Details
      </Button>
    </div>,
  ],
]
`;

exports[`Webhook Component: Batch Status Tab should show empty message when no batches exist 1`] = `
<Empty
  message="There are no batches for your webhook"
/>
`;

exports[`Webhook Component: Batch Status Tab should show loading component while refreshing data 1`] = `
<PanelLoading
  minHeight="400px"
/>
`;
